File: practice/src/main/java/com/decade/practice/security/strategy/LogoutStrategy.java
package com.decade.practice.security.strategy;

import com.decade.practice.core.TokenCredentialService;
import com.decade.practice.utils.TokenUtils;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.springframework.security.core.Authentication;
import org.springframework.security.web.authentication.logout.LogoutHandler;
import org.springframework.stereotype.Service;

@Service
public class LogoutStrategy implements LogoutHandler {

      private final TokenCredentialService credentialService;

      public LogoutStrategy(TokenCredentialService credentialService) {
            this.credentialService = credentialService;
      }

      @Override
      public void logout(
            HttpServletRequest request,
            HttpServletResponse response,
            Authentication authentication
      ) {
            String refreshToken = TokenUtils.extractRefreshToken(request);
            if (refreshToken == null) {
                  return;
            }
            String username = authentication.getName();
            credentialService.evict(username, refreshToken);
      }
}

File: practice/src/main/java/com/decade/practice/security/LogoutStrategy.kt
package com.decade.practice.security

import com.decade.practice.util.TokenUtils
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.security.core.Authentication
import org.springframework.security.web.authentication.logout.LogoutHandler
import org.springframework.stereotype.Service

@Service
class LogoutStrategy(
      private val credentialService: TokenCredentialService
) : LogoutHandler {

      override fun logout(
            request: HttpServletRequest,
            response: HttpServletResponse,
            authentication: Authentication
      ) {
            val refreshToken = TokenUtils.extractRefreshToken(request)
            if (refreshToken == null) {
                  return
            }
            val username = authentication.name
            credentialService.evict(username, refreshToken)
      }
}
